### PR TITLE
feat: optimism l1 fees

### PR DESCRIPTION
**Pull request by kaladinlight**
## Description

- use `maxFeePerGas` as `gasPrice` for chains that support eip1559
- the above change allows us to remove zrx gas api and the average vs fast comparison logic
- add more constants for L1 gas limits to account for L1 fees on optimism (swaps/approvals)
- a little bit of type narrowing jank to access `l1GasPrice`
- remove `totalFee` which looks to be unused and irrelevant (correct me if I am wrong)

The L1 fee edge case is small and was also covered up by the arbitrary 50% gas limit buffer that existed before the eip1559 fee refactor.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

https://discord.com/channels/554694662431178782/876953436908822568/1101532729859911765

## Risk

Low/Med

## Testing

- Test Zrx swap and approval flow on Optimism
- Sanity check defi interaction (fox-farming/univ2) and swaps (no gas limit/broadcast issues)
- The actual edge case will be hard to validate on optimism just do to slim eth on optimism balance window, @gomesalexandre may still be in a wallet state to validate here

### Engineering

:point_up: 

### Operations

:point_up:

### CODE REVIEW

File: packages/chain-adapters/src/evm/avalanche/AvalancheChainAdapter.ts [hunk 1]
```diff
@@ -61,23 +60,21 @@ export class ChainAdapter extends EvmBaseAdapter<KnownChainIds.AvalancheMainnet>
...
   async getGasFeeData(): Promise<GasFeeDataEstimate> {
-    const { gasPrice, fast, average, slow } = await this.api.getGasFees()
-
-    const scalars = { fast: bn(1.2), average: bn(1), slow: bn(0.8) }
+    const { fast, average, slow } = await this.api.getGasFees()
 
     return {
       fast: {
-        gasPrice: calcFee(gasPrice, 'fast', scalars),
+        gasPrice: fast.maxFeePerGas ?? '0',
```
Comment (kaladinlight) [inline packages/chain-adapters/src/evm/avalanche/AvalancheChainAdapter.ts:67]: will be improved with the unchained legacy fee update coming up next, but there should always be `maxFeePerGas` available on chains that support eip1559 (ethereum, avalanche, polygon) even though it is typed as optional

File: packages/chain-adapters/src/evm/avalanche/AvalancheChainAdapter.ts [hunk 2]
```diff
@@ -94,15 +91,15 @@ export class ChainAdapter extends EvmBaseAdapter<KnownChainIds.AvalancheMainnet>
 
     return {
       fast: {
-        txFee: getTxFee({ gasLimit, ...fast }),
+        txFee: bnOrZero(bn(fast.gasPrice).times(gasLimit)).toFixed(0),
```
Comment (kaladinlight) [inline packages/chain-adapters/src/evm/avalanche/AvalancheChainAdapter.ts:94]: since `gasPrice` and `maxFeePerGas` are the same value now, we can simplify the `txFee` calculation

File: packages/chain-adapters/src/evm/ethereum/EthereumChainAdapter.ts [hunk 3]
```diff
@@ -67,26 +60,21 @@ export class ChainAdapter extends EvmBaseAdapter<KnownChainIds.EthereumMainnet>
   }
 
   async getGasFeeData(): Promise<GasFeeDataEstimate> {
-    const { data: responseData } = await axios.get<ZrxGasApiResponse>('https://gas.api.0x.org/')
```
Comment (kaladinlight) [inline packages/chain-adapters/src/evm/ethereum/EthereumChainAdapter.ts:70]: bye bye zrx gas api!!!

File: packages/chain-adapters/src/evm/optimism/OptimismChainAdapter.ts [hunk 4]
```diff
@@ -14,6 +14,10 @@ import type { GasFeeDataEstimate } from '../types'
 const SUPPORTED_CHAIN_IDS = [KnownChainIds.OptimismMainnet]
 const DEFAULT_CHAIN_ID = KnownChainIds.OptimismMainnet
 
+export const isOptimismChainAdapter = (adapter: unknown): adapter is ChainAdapter => {
```
Comment (kaladinlight) [inline packages/chain-adapters/src/evm/optimism/OptimismChainAdapter.ts:17]: added to narrow type and access `l1GasPrice` for zrx

File: src/features/defi/helpers/utils.ts [hunk 5]
```diff
@@ -46,15 +46,11 @@ export const canCoverTxFees = ({
...
-  txFee: BigNumber.max(
-    average.txFee,
-    bnOrZero(bn(fast.chainSpecific.gasPrice).times(average.chainSpecific.gasLimit)),
-  ).toFixed(0), // use worst case average eip1559 vs fast legacy
+  txFee: average.txFee,
   chainSpecific: {
-    gasLimit: average.chainSpecific.gasLimit, // average and fast gasLimit values are the same
-    gasPrice: fast.chainSpecific.gasPrice, // fast gas price since it is underestimated currently
+    gasLimit: average.chainSpecific.gasLimit,
+    gasPrice: average.chainSpecific.gasPrice,
```
Comment (kaladinlight) [inline src/features/defi/helpers/utils.ts:53]: we can just use average across the board now since it will be a viable `gasPrice` instead of the low balled zrx gas estimate

File: src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts [hunk 6]
```diff
@@ -63,19 +69,34 @@ export async function getZrxTradeQuote<T extends ZrxSupportedChainId>(
...
+      if (optimism.isOptimismChainAdapter(adapter)) {
+        const { average, l1GasPrice } = await adapter.getGasFeeData()
 
-    const useSellAmount = !!sellAmount
-    const rate = useSellAmount ? data.price : bn(1).div(data.price).toString()
-    const gasLimit = bnOrZero(data.gas)
-    const txFee = gasLimit.times(maxGasPrice)
+        // account for l1 transaction fees for optimism
+        const l1ApprovalFee = bn(OPTIMISM_L1_APPROVE_GAS_LIMIT).times(l1GasPrice)
+        const l1TxFee = bn(OPTIMISM_L1_SWAP_GAS_LIMIT).times(l1GasPrice)
```
Comment (kaladinlight) [inline src/lib/swapper/swappers/ZrxSwapper/getZrxTradeQuote/getZrxTradeQuote.ts:84]: Add safe l1 gas limit fee buffer to optimism transactions so balance check works and we don't end up with failed transactions down the line

**Review by gomesalexandre — COMMENTED** [review]
Tested locally, confirmed this properly handles the too low for L1 fee, just enough for Optimism fee edge case:

image

Additionally, regression tested ZRX swaps on mainnet, unfortunately, this seems to have borked maxes, see this diff vs. prod:

image
image

image
image

**Review by gomesalexandre — APPROVED** [review]
Retested locally, confirmed:

- This still handles too low L1 fees

image

- Latest commits unrug ZRX max swaps with the buffer and quote brought back

image
